smart-contract-extensions/contracts/KeyPurchaserFactory.sol

```diff
@@ -0,0 +1,106 @@
+pragma solidity 0.5.17;
+
+import './KeyPurchaser.sol';
+import 'hardlydifficult-ethereum-contracts/contracts/proxies/Clone2Factory.sol';
+import 'hardlydifficult-ethereum-contracts/contracts/proxies/Clone2Probe.sol';
+import 'unlock-abi-7/IPublicLockV7.sol';
+
+/**
+ * @notice A factory for creating keyPurchasers.
+ * @dev This contract acts as a registry to discover purchasers for a lock
+ * and by creating each purchaser itself, it's a single tx to deploy and initialize and this guarantees
+ * a consistent implementation and that it was created by the lock owner.
+ */
+contract KeyPurchaserFactory
+{
+  using Clone2Factory for address;
+  using Clone2Probe for address;
+
+  event KeyPurchaserCreated(address indexed forLock, address indexed keyPurchaser);
+
+  /**
+   * @notice The implementation for minimal proxies to use.
+   */
+  address public keyPurchaserTemplate;
+
+  /**
+   * @notice The registry of available keyPurchases for a given lock.
+   */
+  mapping(address => address[]) public lockToKeyPurchasers;
+
+  constructor() public
+  {
+    keyPurchaserTemplate = address(new KeyPurchaser());
+  }
+
+  function _getClone2Salt(
+    IPublicLockV7 _lock,
+    uint _maxKeyPrice,
+    uint _renewWindow,
+    uint _renewMinFrequency,
+    uint _msgSenderReward
+  ) private pure
+    returns (bytes32)
+  {
+    return keccak256(abi.encodePacked(address(_lock), _maxKeyPrice, _renewWindow, _renewMinFrequency, _msgSenderReward));
+  }
+
+  /**
+   * @notice Returns the address for a keyPurchaser given the terms.
+   * @dev This works whether or not the keyPurchaser has actually been deployed yet.
+   */
+  function getExpectedAddress(
+    IPublicLockV7 _lock,
+    uint _maxKeyPrice,
+    uint _renewWindow,
+    uint _renewMinFrequency,
+    uint _msgSenderReward
+  ) external view
+    returns (address)
+  {
+    bytes32 salt = _getClone2Salt(_lock, _maxKeyPrice, _renewWindow, _renewMinFrequency, _msgSenderReward);
+    return keyPurchaserTemplate.getClone2Address(salt);
+  }
+
+  /**
+   * @notice Deploys a new KeyPurchaser for a lock and stores the address for reference.
+   */
+  function deployKeyPurchaser(
+    IPublicLockV7 _lock,
+    uint _maxKeyPrice,
+    uint _renewWindow,
+    uint _renewMinFrequency,
```

Comment:
So for these 2 params:
`_renewWindow` is to specify at what point in time a key is eligible for renewal, and
`_renewMinFrequency` is to specify the min. renewal period.. ex: every year, or every month, etc..
Do I have this straight?

Reply:
Just saw the comments in the first contract.